Peer Code Review - Guidance Note (#44)
worldbank/dime-standards: dime-coding-standards/checklists/tex/peer_code_review_guidance_note.tex

\titleBox{
	\textcolor{white}{\LARGE{\textbf{DIME Analytics \\ Quarterly Peer Code Review Guidance Note}} \\
	\Large\textbf{{v1.0 - Last updated November 18, 2022}}}
}

	\section*{Overview}
 Every quarter, DIME Analytics organizes the \textbf{Peer Code Review}, which is a real-time code and data-quality assurance process. This is a structured opportunity for participants to exchange, run, and provide feedback on each other’s code. Apart from increasing reproducibility, transparency and adherence to best practices, it is a great learning opportunity - code writers get to look at their work from a different perspective, and reviewers are exposed to different styles and practices. 
\hfill \break
\begin{flushleft} The peer code review is designed for scripts that are modular enough to be understood on their own (a reviewer can understand the script without having to refer to other project code files) and are short enough for a reviewer to read through in approximately half a day (depending on complexity). We recommend submitting recently-completed tasks, to allow corrections to be made in real time.

\section*{How it works}

	\begin{enumerate}
	    
	  
		\setlength\itemsep{-0.1em}

> 		\item Teams \href{https://survey.wb.surveycto.com/collect/code_review_sign_up?caseid=}{sign-up} to participate in the code review each quarter. All DIME Research Assistants working on code are expected to participate, and any Bank staff or consultant who would like to get feedback on their code is welcome to participate.

Reviewer: We should think about this link. as it will be public we should make sure sign-up is off except during the designated sign-up periods, or make sure that the dates for the next round are clear. like it wouldn't be a problem if someone signs up from another unit now if they know the round they are signing up for doesn't take place until February. maybe ideal is that in the periods in-between sign-ups there's just a message that says very clearly that the next round won't be until X date and that sign-up will begin at Y date.

    

		\item All participants prepare a code package to submit to review, following \href{https://github.com/worldbank/dime-standards/blob/master/dime-coding-standards/checklists/Peer%20Code%20Review%20Submission%20Checklist.pdf}{these guidelines}. Submitted code should be \textbf{no longer than 1000 lines}. 
        
        \begin{itemize}

>             \item As part of this step, teams should evaluate if they would like to receive comments on \textbf{code reproducibility} and \textbf{output replicability}. 

Reviewer: ```suggestion
            \item As part of this step, teams will identify what tasks they want review (data cleaning, construction, or analysis) and whether they want the reviewer to assess  \textbf{computational reproducibility}. 
```

>             \item \textbf{Note} that in this case, teams should prepare a \textbf{de-identified} version of the dataset to include in the submission package.

Reviewer: ```suggestion
            \item \textbf{Note} In order for the reviewer to assess reproducibility, the peer review submission package must include a \textbf{de-identified} version of the dataset.
```

        \end{itemize}

>         \item Participants are then paired for review based on preferred statistical software and length of code.

Reviewer: ```suggestion
        \item Participants are paired for review based on preferred statistical software and length of code.
```

        \item The actual peer code review activity takes place over the course of one week, which includes:
         \begin{itemize}
            \item A \textbf{kickoff session} with an overview of the process, and pair assignments. 

>             \item A \textbf{group work party} with technical support from DIME Analytics. Pairs work together, and answer questions to allow each other to review their submitted code.

Reviewer: ```suggestion
            \item A \textbf{group work session} with technical support from DIME Analytics. Pairs work together, and answer questions to allow each other to review their submitted code.
```

        \end{itemize}

>         \item Reviewers finally submit feedback using \href{https://survey.wb.surveycto.com/collect/code_review_summary?caseid=}{this form}. They also fill out and submit the following checklists:

Reviewer: ```suggestion
       \item Reviewers review the code using structured checklists:
```

        \begin{itemize}
            \item \textbf{Mandatory}: \href{https://github.com/worldbank/dime-standards/blob/master/dime-coding-standards/checklists/Reviewer%20Feedback%20Checklist.pdf}{Reviewer Feedback Checklist}
            \item \textbf{Depending on tasks in submitted code:} 
            \begin{itemize}
                \item \href{https://github.com/worldbank/dime-standards/blob/master/dime-coding-standards/checklists/Cleaning%20Code%20Review%20Checklist.pdf}{Cleaning Checklist}
                \item \href{https://github.com/worldbank/dime-standards/blob/master/dime-coding-standards/checklists/Construction%20Code%20Review%20Checklist.pdf}{Construction Checklist}
                \item \href{https://github.com/worldbank/dime-standards/blob/master/dime-coding-standards/checklists/Analysis%20Code%20Review%20Checklist.pdf}{Analysis Checklist}

>             \end{itemize}

Reviewer: INSERT:          
\item At the end of the week, reviewers submit \href{https://survey.wb.surveycto.com/collect/code_review_summary?caseid=}{a structured feedback form} and upload their completed checklists.

        \end{itemize}

        \end{enumerate}
	
	\section{What teams receive}
        \begin{itemize}
 

>         \item \textbf{TTLs} receive \href{https://github.com/worldbank/dime-standards/blob/peer-code-review/dime-coding-standards/checklists/Sample%20TTL%20Report.pdf}{individual reports} for their projects indicating adherence to best practices, and areas for improvement.

Reviewer: ```suggestion
        \item \textbf{TTLs} receive \href{https://github.com/worldbank/dime-standards/blob/peer-code-review/dime-coding-standards/checklists/Sample%20TTL%20Report.pdf}{a standardized report} for their project indicating adherence to best practices and areas for improvement.
```

>         \item \textbf{Research assistants} receive detailed feedback in the form of the \href{https://github.com/worldbank/dime-standards/blob/master/dime-coding-standards/checklists/Reviewer%20Feedback%20Checklist.pdf}{Reviewer Feedback Checklist} submitted by their code review partners.

Reviewer: ```suggestion
        \item \textbf{Research assistants} receive detailed feedback though the \href{https://github.com/worldbank/dime-standards/blob/master/dime-coding-standards/checklists/Reviewer%20Feedback%20Checklist.pdf}{Reviewer Feedback Checklist} submitted by their review partner.
```

>         \item \textbf{DIME Analytics} prepares and shares a \href{https://github.com/worldbank/dime-standards/blob/peer-code-review/dime-coding-standards/checklists/Sample%20Overall%20Summary%20Report%20.pdf}{summary report}  for the completed round of peer code review with DIME management, as well as the teams who signed up for the code review round.

Reviewer: ```suggestion
        \item All participating \textbf{TTLs and DIME Managers} receive a \href{https://github.com/worldbank/dime-standards/blob/peer-code-review/dime-coding-standards/checklists/Sample%20Overall%20Summary%20Report%20.pdf}{summary report} highlighting common strengths and weaknesses.
```

\hfill \break
\begin{flushleft} The peer code review is designed for scripts that are modular enough to be understood on their own (a reviewer can understand the script without having to refer to other project code files) and are short enough for a reviewer to read through in approximately half a day (depending on complexity). We recommend submitting recently-completed tasks, to allow corrections to be made in real time.

\section*{How it works}

	\begin{enumerate}
	    
	  
		\setlength\itemsep{-0.1em}
		\item Teams \href{https://survey.wb.surveycto.com/collect/code_review_sign_up?caseid=}{sign-up} to participate in the code review each quarter. All DIME Research Assistants working on code are expected to participate, and any Bank staff or consultant who would like to get feedback on their code is welcome to participate.
    

		\item All participants prepare a code package to submit to review, following \href{https://github.com/worldbank/dime-standards/blob/master/dime-coding-standards/checklists/Peer%20Code%20Review%20Submission%20Checklist.pdf}{these guidelines}. Submitted code should be \textbf{no longer than 1000 lines}. 
        
        \begin{itemize}

>             \item As part of this step, teams will identify what tasks they want review (data cleaning, construction, or analysis) and whether they want the reviewer to assess  \textbf{computational reproducibility}. 

Reviewer: ```suggestion
            \item As part of this step, teams will identify what tasks they want reviewed (data cleaning, construction, or analysis) and whether they want the reviewer to assess  \textbf{computational reproducibility}. 
```

            \item \textbf{Note} In order for the reviewer to assess reproducibility, the peer review submission package must include a \textbf{de-identified} version of the dataset.
        \end{itemize}

        \item Participants are paired for review based on preferred statistical software and length of code.

        \item The actual peer code review activity takes place over the course of one week, which includes:
         \begin{itemize}
            \item A \textbf{kickoff session} with an overview of the process, and pair assignments. 
            \item A \textbf{group work session} with technical support from DIME Analytics. Pairs work together, and answer questions to allow each other to review their submitted code.
        \end{itemize}

       \item Reviewers review the code using structured checklists:
        \begin{itemize}

>             \item \textbf{Mandatory}: \href{https://github.com/worldbank/dime-standards/blob/master/dime-coding-standards/checklists/Reviewer%20Feedback%20Checklist.pdf}{Reviewer Feedback Checklist}

Reviewer: Given that reproducibility check isn't mandatory, i think we want to remove that from the reviewer feedback sheet and create a separate checklist (the same way we have cleaning, construction, etc).

Author: That sounds good, I'll make those changes to the checklists. And I also agree on the point about re-organizing the folder structure. I'll create a new one for this.

            \item \textbf{Depending on tasks in submitted code:} 
            \begin{itemize}
                \item \href{https://github.com/worldbank/dime-standards/blob/master/dime-coding-standards/checklists/Cleaning%20Code%20Review%20Checklist.pdf}{Cleaning Checklist}
                \item \href{https://github.com/worldbank/dime-standards/blob/master/dime-coding-standards/checklists/Construction%20Code%20Review%20Checklist.pdf}{Construction Checklist}
                \item \href{https://github.com/worldbank/dime-standards/blob/master/dime-coding-standards/checklists/Analysis%20Code%20Review%20Checklist.pdf}{Analysis Checklist}
            \end{itemize}
        \end{itemize}
        
         \item At the end of the week, reviewers submit \href{https://survey.wb.surveycto.com/collect/code_review_summary?caseid=}{a structured feedback form} and upload their completed checklists.

>          \item \textbf{DIME Analytics} conducts a \textbf{code review summary session} \textbf{(within 3 weeks)} with RAs, as well as TTLs (optional), to discuss summary statistics and general lessons from the recently concluded round of peer code review.

Reviewer: ```suggestion
         \item \textbf{DIME Analytics} conducts a \textbf{code review summary session} \textbf{(within 3 weeks)} with participants, to discuss summary statistics and general lessons from the recently concluded round of peer code review. TTLs are invited to attend.
```
